### PR TITLE
fix: Update scheduler script

### PR DESCRIPTION
**Pull request by ivadym**
## Description

Refactors the scheduler script and removes the qsub option, currently non-functional in production. If we choose to reintroduce it in the future, the process will be straightforward as the code and models are easily extendable.

This guide explains the need for overwriting the Snakemake scheduler and could help in understanding the code execution flow: https://r3.pages.uni.lu/school/snakemake-tutorial/HPC#immediate_submit

#### Changed

- Cluster scheduler script for immediate submit

#### Removed

- SGE support

## Documentation

- [ ] N/A
- [x] Updated Balsamic documentation to reflect the changes as needed for this PR.
    - `user_guide.rst`

### Feature Tests

- [ ] N/A
- [x] Test successful job submission 
    - [x] TGA T/O case
        - ~ Submission* time this PR: `3m2.062s`,  
        - ~ Submission* time master: `2m23.637s`       
    - [x] WGS T/N case
        - ~ Submission* time this PR: `3m48.636s`,
        - ~ Submission* time master: `2m47.884s`
    - [x] Manual submission Balsamic command
- [x] Test correct generation of the `slurm_jobids.yaml` file (for job parsing by trailblazer)
    Screenshot 2024-02-02 at 14 17 12

* `cg workflow balsamic start <case_id>`

### Pipeline Integrity Tests

- **Report deliver (generation of the `.hk` file)**
    - [x] N/A
    - [ ] Verified
- **TGA T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **TGA T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **UMI T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **UMI T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **WGS T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **WGS T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **QC Workflow**
    - [x] N/A
    - [ ] Verified
- **PON Workflow**
    - [x] N/A
    - [ ] Verified

## Clinical Genomics Stockholm

<details>
<summary></summary>

### Documentation

- **Atlas documentation**
    - [x] N/A
    - [ ] Updated: [Link]
- **Web portal for Clinical Genomics**
    - [x] N/A
    - [ ] Updated: [Link]

### User Changes

- [x] N/A
- [ ] This PR affects the output files or results.
    - [ ] User feedback is considered unnecessary because [Justification].
    - [ ] Affected users have been included in the development process and given a chance to provide feedback.

### Infrastructure Changes

- **Stored files in Housekeeper**
    - [x] N/A
    - [ ] Updated: [Link]
- **CG (CLI and delivered/uploaded files)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Servers (configuration files on Hasta)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Scout interface**
    - [x] N/A
    - [ ] Updated: [Link]

### Integration Tests

- [x] N/A
- [ ] Test [Description]
    - [Screenshot]

</details>

## Checklist

> [!IMPORTANT]  
> Ensure that all checkboxes below are ticked before merging.

### For Developers

- **PR Description**
    - [x] Provided a comprehensive description of the PR.
    - [x] Linked relevant user stories or issues to the PR.
- **Documentation**
    - [x] Verified and updated documentation if necessary.
- **Tests**
    - [x] Described and tested the functionality addressed in the PR.
    - [x] Ensured integration of the new code with existing workflows.
    - [x] Confirmed that meaningful unit tests were added for the changes introduced.
    - [x] Checked that the PR has successfully passed all relevant code smells and coverage checks.
- **Review**
    - [x] Addressed and resolved all the feedback provided during the code review process.
    - [x] Obtained final approval from designated reviewers.

### For Reviewers

- **Code**
    - [ ] Code implements the intended features or fixes the reported issue.
    - [ ] Code follows the project's coding standards and style guide.
- **Documentation**
    - [ ] Pipeline changes are well-documented in the CHANGELOG and relevant documentation.
- **Tests**
    - [ ] The author provided a description of their manual testing, including consideration of edge cases and boundary
      conditions where applicable, with satisfactory results.
- **Review**
    - [ ] Confirmed that the developer has addressed all the comments during the code review.

### CODE REVIEW

File: BALSAMIC/assets/scripts/immediate_submit.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    OPTION_CLUSTER_MAIL,
+    OPTION_CLUSTER_MAIL_TYPE,
+    OPTION_CLUSTER_PROFILE,
+    OPTION_CLUSTER_QOS,
+)
+from BALSAMIC.constants.cluster import QOS, ClusterProfile
+from BALSAMIC.models.scheduler import Scheduler
+
+
+@click.command()
```
Comment (mathiasbio) [inline BALSAMIC/assets/scripts/immediate_submit.py:20]: Wow, this CLI surprised me, is there a use-case you have in mind for this where it will be called outside the regular balsamic run?

File: BALSAMIC/commands/run/analysis.py [hunk 2]
```diff
@@ -92,9 +92,7 @@ def analysis(
         run_mode: RunMode = RunMode.LOCAL
 
     if run_mode == RunMode.CLUSTER and not account:
-        LOG.info(
-            "slurm-account, qsub-account, or account is required for slurm run mode"
-        )
+        LOG.info("An account is required for custer run mode")
```
Comment (mathiasbio) [inline BALSAMIC/commands/run/analysis.py:95]:
```suggestion
        LOG.info("An account is required for cluster run mode")
```

File: BALSAMIC/assets/scripts/immediate_submit.py (hunk 1 above)
Comment (mathiasbio) [reply BALSAMIC/assets/scripts/immediate_submit.py:20]: Oh I see, this CLI is used by the balsamic run? I realize I have no idea how these jobs are submitted

**Comment by mathiasbio** [comment]
Can I ask, beyond the refactoring and improvements in the code, are there any other expected effects from this change on the practical running of the analyses? For instance do we expect that it will affect the key_error in some way?

**Comment by ivadym** [comment]
> Can I ask, beyond the refactoring and improvements in the code, are there any other expected effects from this change on the practical running of the analyses? For instance do we expect that it will affect the key_error in some way?

I have just run it once and the submission failed due to a key error so it doesn't look good... I hoped it was something to do with the scheduler but seems more and more a bug in how Snakemake handles finished jobs for immediate submit

File: BALSAMIC/assets/scripts/immediate_submit.py (hunk 1 above)
Comment (ivadym) [reply BALSAMIC/assets/scripts/immediate_submit.py:20]: Exactly, it is provided as a script to the Snakemake command (`--cluster` option)

File: BALSAMIC/constants/workflow_params.py [hunk 3]
```diff
@@ -115,7 +115,7 @@
     },
 }
 
-SLEEP_BEFORE_START = 300
+SLEEP_BEFORE_START = 600
```
Comment (mathiasbio) [inline BALSAMIC/constants/workflow_params.py:118]: 🌟

File: BALSAMIC/models/scheduler.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        return ""
+
+    def get_acctg_freq_option(self) -> str:
+        """Return string representation of the profiling sampling intervals in seconds option."""
+        if self.benchmark and self.profile == ClusterProfile.SLURM:
+            return f"--acctg-freq {self.profiling_type}={self.profiling_interval}"
+        return ""
+
+    def get_ntasks_option(self) -> str:
+        """Return the maximum of number tasks for allocation."""
```
Comment (mathiasbio) [inline BALSAMIC/models/scheduler.py:117]:
```suggestion
        """Return the maximum number of tasks for allocation."""
```

File: CHANGELOG.rst [hunk 5]
```diff
@@ -1,8 +1,15 @@
+[X.X.X]
+
+Changed:
+^^^^^^^^
+* Cluster scheduler script for immediate submit https://github.com/Clinical-Genomics/BALSAMIC/pull/1372
+* `SLEEP_BEFORE_START` to 600s
```
Comment (mathiasbio) [inline CHANGELOG.rst:6]:
```suggestion
* `SLEEP_BEFORE_START` to 600s https://github.com/Clinical-Genomics/BALSAMIC/pull/1372
```

**Comment by mathiasbio** [comment]
> > Can I ask, beyond the refactoring and improvements in the code, are there any other expected effects from this change on the practical running of the analyses? For instance do we expect that it will affect the key_error in some way?
> 
> I have just run it once and the submission failed due to a key error so it doesn't look good... I hoped it was something to do with the scheduler but seems more and more a bug in how Snakemake handles finished jobs for immediate submit

How has this been resolved? Is it just that we increased the sleep? I

**Review by mathiasbio — APPROVED** [review]
Amazing effort 🌟 👍 👍 

Do you think there will be any effects on production from the increased submission times in this update? Will it decrease the need for sleep time because it's slower?

**Comment by ivadym** [comment]
> > > Can I ask, beyond the refactoring and improvements in the code, are there any other expected effects from this change on the practical running of the analyses? For instance do we expect that it will affect the key_error in some way?
> > 
> > 
> > I have just run it once and the submission failed due to a key error so it doesn't look good... I hoped it was something to do with the scheduler but seems more and more a bug in how Snakemake handles finished jobs for immediate submit
> 
> How has this been resolved? Is it just that we increased the sleep? : )

Unfortunately nothing has changed regarding the Key Error :(

I think if we want to get rid of it we would need to fork snakemake and catch the Key Error

**Comment by mathiasbio** [comment]
> > Amazing effort 🌟 👍 👍
> > Do you think there will be any effects on production from the increased submission times in this update? Will it decrease the need for sleep time because it's slower?
> 
> I think it could make it worse. Slower submission times, even by just around one minute, will increase the chances of encountering a Key Error. Should be OK since we also increased the sleep time but I will test it once more to make sure the submission times align with what I have tested before.
> 
> We can scrap this PR, but if we would like to add memory allocation in the future this PR would make it way easier

I'm in favour of deploying it, I don't think adding a few more minutes to the sleep rule would be an issue anyway! Probably better to be able to add memory allocation!

**Comment by ivadym** [comment]
I've tested it again, same results as the ones I included before
